src/bank.rs

```diff
@@ -349,36 +359,61 @@ impl Bank {
         let mut ret = Vec::new();
         for (i, id) in ids.iter().enumerate() {
             if let Some(entry) = last_ids.sigs.get(id) {
-                if ((last_ids.nth - entry.2) as usize) <= MAX_ENTRY_IDS {
+                if ((last_ids.nth - entry.2) as usize) < MAX_ENTRY_IDS {
                     ret.push((i, entry.1));
                 }
             }
         }
         ret
     }
 
+    /// Looks through a list of tick heights and stakes, and finds the latest
+    /// tick that has achieved finality
+    pub fn get_finality_timestamp(
+        &self,
+        ticks_and_stakes: &mut [(u64, i64)],
+        supermajority_stake: i64,
+    ) -> Option<u64> {
+        // Sort by tick height
+        ticks_and_stakes.sort_by(|a, b| a.0.cmp(&b.0));
+        let last_ids = self.last_ids.read().unwrap();
+        let nth = last_ids.nth;
+        let mut total = 0;
+        for (tick_height, stake) in ticks_and_stakes.iter() {
+            if ((nth - tick_height) as usize) < MAX_ENTRY_IDS {
+                total += stake;
+                if total > supermajority_stake {
+                    return Self::tick_height_to_timestamp(&*last_ids, *tick_height);
+                }
+            }
+        }
+        None
+    }
+
     /// Tell the bank which Entry IDs exist on the ledger. This function
     /// assumes subsequent calls correspond to later entries, and will boot
     /// the oldest ones once its internal cache is full. Once boot, the
     /// bank will reject transactions using that `last_id`.
     pub fn register_entry_id(&self, last_id: &Hash) {
         let mut last_ids = self.last_ids.write().unwrap();
 
+        // Increment "nth" so that the index of this tick in the LastIds structure
+        // matches the "tick height" when the validator votes for this tick
+        last_ids.nth += 1;
```

Comment:
if last_ids.nth and tick_height don't match, what then?

Reply:
When a validator submits a vote for a particular tick, the height in that tick should always match the second u64 in the tuple of the "sigs" hashmap (HashMap<Hash, (SignatureStatusMap, u64, u64)>), because the ticks are replayed in the same order on the validator as they were generated on the leader.

Reply:
what I'm getting at is that last_id is currently a tick, but that coupling is new, not a hard one.  there are tests that introduce last_ids that aren't ticks.